### PR TITLE
Site Settings: Update the general tab to use toggles instead of checkboxes

### PR DESCRIPTION
**Pull request by tyxla**
### Description

This PR:
* Refactors the existing general settings to use toggles instead of checkboxes (see #9171). 
* Removes the `dirtyLinkedState` mixin dependency. 
* Fixes a notice in `<Timezone />` when no value is specified.
* Implements `i18n-calypso.localize` for translated strings.

Inspired by @ryelle's awesome work in #10167.

### Screenshots
**Jetpack Site**

**WordPress.com site**

### To test

* Get this branch going locally or on calypso.live.
* Visit the General settings at /settings/general/$site for a Jetpack site.
* Make any changes.
* Verify that all changes have been saved and applied to your Jetpack site.

Also, test on WordPress.com sites (it contains **Language** and **Site Timezone** options, as well as a **Private** option for the **Privacy** setting):

* Get this branch going locally or on calypso.live.
* Visit the General settings at /settings/general/$site for a WordPress.com site.
* Make any changes.
* Verify that all changes have been saved and applied to your WordPress.com site (you can also check the production settings page).

/cc 
@ryelle @oskosk @roccotripaldi @johnHackworth for code review 
@rickybanister @MichaelArestad for design review

### CODE REVIEW

**Comment by rickybanister** [comment]
Looks good to me.

**Comment by rickybanister** [comment]
I'll expand on that a bit—we're rightful in using toggles in situations where the choice is on/off binary (as it is in all of these cases). Let's just be sure to avoid swapping any 'multiple choice' checkboxes for toggles. Carry on!

**Comment by MichaelArestad** [comment]
This looks spot on. Nice work!

File: client/my-sites/site-settings/form-general.jsx [hunk 1]
```diff
@@ -127,35 +120,68 @@ const FormGeneral = React.createClass( {
...
+
+	handleChange( event ) {
+		const currentTargetName = event.currentTarget.name,
+			currentTargetValue = event.currentTarget.value;
+
+		this.setDirtyField( currentTargetName );
+		this.setState( { [ currentTargetName ]: currentTargetValue } );
+	},
+
+	handleToggle( name ) {
```
Comment (ryelle) [inline client/my-sites/site-settings/form-general.jsx]: I wonder if we should move these functions into `form-base`, if we're going to have the same functions across a few tabs. Perhaps another PR after this & the discussion tab is merged.

File: client/my-sites/site-settings/form-general.jsx [hunk 2]
```diff
@@ -509,39 +545,40 @@ const FormGeneral = React.createClass( {
...
-				<legend>{ this.translate( 'Holiday Snow' ) }</legend>
-				<ul>
-					<li>
-						<FormLabel>
-							<FormCheckbox name="holidaysnow" checkedLink={ this.linkState( 'holidaysnow' ) }/>
-							<span>{ this.translate( 'Show falling snow on my blog until January 4th.' ) }</span>
-						</FormLabel>
-					</li>
-				</ul>
+				<legend>{ translate( 'Holiday Snow' ) }</legend>
```
Comment (ryelle) [inline client/my-sites/site-settings/form-general.jsx]: You could also use `FormLegend`, which adds the `form-legend` class.

File: client/my-sites/site-settings/form-general.jsx [hunk 3]
```diff
@@ -509,39 +545,40 @@ const FormGeneral = React.createClass( {
...
 			<FormFieldset>
 				<FormLabel htmlFor="blogtimezone">
-					{ this.translate( 'Site Timezone' ) }
+					{ translate( 'Site Timezone' ) }
 				</FormLabel>
 
 				<Timezone
-					valueLink={ this.linkState( 'timezone_string' ) }
-					selectedZone={ this.linkState( 'timezone_string' ).value }
+					value={ this.state.timezone_string }
```
Comment (ryelle) [inline client/my-sites/site-settings/form-general.jsx]: Is `value` used in this component? It looks like only `selectedZone` is used. I think even the `valueLink` that was here was ignored.

File: client/my-sites/site-settings/form-general.jsx [hunk 4]
```diff
@@ -260,12 +290,12 @@ const FormGeneral = React.createClass( {
 						name="blog_public"
 						value="1"
 						checked={ 1 === parseInt( this.state.blog_public, 10 ) }
-						onChange={ this.handleRadio }
+						onChange={ this.handleChange }
 						disabled={ this.state.fetchingSettings }
 						onClick={ this.onRecordEvent( 'Clicked Site Visibility Radio Button' ) } />
-					<span>{ this.translate( 'Public' ) }</span>
+					<span>{ translate( 'Public' ) }</span>
 					<FormSettingExplanation isIndented>
```
Comment (ryelle) [inline client/my-sites/site-settings/form-general.jsx]: Should this be indented? @MichaelArestad the grey text under each option

screen shot 2016-12-21 at 11 30 14 am

File: client/my-sites/site-settings/style.scss [hunk 5]
```diff
@@ -160,6 +160,26 @@
...
+	.form-toggle__switch {
+		flex: 1 0 auto;
+	}
+
+	.form-toggle__switch + span {
+		flex: 0 1 100%;
+		margin-left: 12px;
+	}
+
+	p.form-setting-explanation.is-indented {
```
Comment (ryelle) [inline client/my-sites/site-settings/style.scss]: Regardless of whether `FormSettingExplanation` should be indented, this CSS should be removed. If you do end up needing to remove the indent, remove the `isIndented` property from the component.

File: client/my-sites/site-settings/form-general.jsx [hunk 6]
```diff
@@ -127,35 +120,68 @@ const FormGeneral = React.createClass( {
 		return this.recordEventOnce.bind( this, key, eventAction );
 	},
 
+	setDirtyField( key ) {
+		const newState = {};
+		const dirtyFields = this.state.dirtyFields || [];
+		if ( dirtyFields.indexOf( key ) === -1 ) {
+			newState.dirtyFields = [ ...dirtyFields, key ];
+		}
+		this.setState( newState );
```
Comment (omarjackman) [inline client/my-sites/site-settings/form-general.jsx]: Updating state here with no changes will still trigger a render. If the intention isn't specifically to do that then I think we should just return early
```js
const dirtyFields = this.state.dirtyFields || [];
if ( dirtyFields.indexOf( key ) >= 0 ) {
    return;
}
this.setState( { dirtyFields: [ ...dirtyFields, key ] } );
```

File: client/my-sites/site-settings/form-general.jsx [hunk 7]
```diff
@@ -127,35 +120,68 @@ const FormGeneral = React.createClass( {
...
+		const currentTargetName = event.currentTarget.name,
+			currentTargetValue = event.currentTarget.value;
+
+		this.setDirtyField( currentTargetName );
+		this.setState( { [ currentTargetName ]: currentTargetValue } );
+	},
+
+	handleToggle( name ) {
+		return () => {
+			this.recordEvent.bind( this, `Toggled ${ name }` );
```
Comment (omarjackman) [inline client/my-sites/site-settings/form-general.jsx]: This line doesn't look like its doing anything. Did we just want to call it?
```js
this.recordEvent( `Toggled ${ name }` );
```

File: client/my-sites/site-settings/form-general.jsx (hunk 1 above)
Comment (omarjackman) [inline client/my-sites/site-settings/form-general.jsx]: Id also consider renaming this to `getToggleHandler`.

File: client/my-sites/site-settings/style.scss (hunk 5 above)
Comment (omarjackman) [inline client/my-sites/site-settings/style.scss]: Also, we probably shouldn't target the `p` tag directly

File: client/my-sites/site-settings/style.scss [hunk 8]
```diff
@@ -160,6 +160,26 @@
...
+	.form-toggle__label {
+		display: flex;
+		align-items: center;
+	}
+
+	.form-toggle__switch {
+		flex: 1 0 auto;
+	}
+
+	.form-toggle__switch + span {
```
Comment (omarjackman) [inline client/my-sites/site-settings/style.scss]: We should give this `span` a class and target it that way if at all possible.

File: client/my-sites/site-settings/form-general.jsx (hunk 6 above)
Comment (omarjackman) [inline client/my-sites/site-settings/form-general.jsx]: Also just noticed that instead of doing  `const dirtyFields = this.state.dirtyFields || [];` we could just initialize the state with an empty array in `getInitialState` and then change this to 
`const { dirtyFields } = this.state;`

File: client/my-sites/site-settings/form-general.jsx (hunk 1 above)
Comment (tyxla) [inline client/my-sites/site-settings/form-general.jsx]: @ryelle I was thinking the same thing! We should definitely do that, but let's do it in another PR after we have both these merged.

@omarjackman I'm fine both ways, but just so you know I've reused the same code that was implemented by @ryelle in #10167. Once we move these into `form-base`, we can consider renaming them.

File: client/my-sites/site-settings/form-general.jsx (hunk 4 above)
Comment (tyxla) [inline client/my-sites/site-settings/form-general.jsx]: If we strive to be consistent, I'd rather keep it non-indented because of how it is already in the Discussion tab in #10167 - see how the "These settings may be overriden..." text appears, and it's not indented. Anyway, let's also see what @MichaelArestad has to say about it.

> Can we ensure that all three types of inputs fix within 24px and that text labels are always exactly 24px (or whatever it is) from the left justification?

Sure, but I think that for Radio buttons it would create a huge white gap. Mini toggles are 24px wide, and have 21px of space to the right, which totals 45px. Radio buttons (together with white space) are equal to 21px. If we add another 24px to the right, the white gap would be too big IMO.

File: client/my-sites/site-settings/form-general.jsx (hunk 3 above)
Comment (tyxla) [inline client/my-sites/site-settings/form-general.jsx]: You're right! :+1:

File: client/my-sites/site-settings/style.scss (hunk 5 above)
Comment (tyxla) [inline client/my-sites/site-settings/style.scss]: @ryelle of course, I've removed it now.

File: client/my-sites/site-settings/form-general.jsx (hunk 6 above)
Comment (tyxla) [inline client/my-sites/site-settings/form-general.jsx]: You're bringing some good points here. Let's see what @ryelle's opinion about that (as I've borrowed this from her code in #10167), and I can later address them in this PR as well.

File: client/my-sites/site-settings/form-general.jsx (hunk 7 above)
Comment (tyxla) [inline client/my-sites/site-settings/form-general.jsx]: Again, this has been borrowed from #10167, but it makes sense. I've removed it from this PR, but @ryelle might want to apply it to #10167 too.

File: client/my-sites/site-settings/style.scss (hunk 8 above)
Comment (tyxla) [inline client/my-sites/site-settings/style.scss]: Good point, I've added a class and styled it by that class. @ryelle you might want to apply this to #10167 as well.

**Comment by tyxla** [comment]
@omarjackman @ryelle @rickybanister @MichaelArestad I've done some updates regarding the comments. This is ready for a new code and design review.

**Comment by tyxla** [comment]
A lot of the work (+ much more) has been addressed in #9085 which as already been merged. 

It'll be a pain to merge all of these together, so I'll start a new PR for the toggle changes.
